[pushbullet] Pushbullet Binding initial contribution

hakan42 opened the pull request:
This is a reimplementation of the pushbullet action which I had authored for openHAB 1:

https://www.openhab.org/addons/actions/pushbullet/

Most of the "business" code is transplanted from OH1, and some adaptations made to be able to use this as an annotated action.

I have not tried to configure the rule via Paper UI, though.

davidgraeff commented on bundles/org.openhab.binding.pushbullet/.classpath:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
```
This looks like an old .classpath file. Was this generated by the eclipse in the new buildsystem environment? (Also have a look at classpath files in other bundles)

hakan42 replied:
the classpath file was a copy from some other binding as I use IntelliJ instead of Eclipse. Could you point me to a "good" one?

davidgraeff replied:
Perfect. @openhab/2-x-add-ons-maintainers Can we please remove all .project and .classpath files now finally. We have the first incidents of a non Eclipse contributor :)

wborn replied:
I don't see why we need to break everyone's workspace by this WIP PR? It's also no first because the Google TTS add-on a few issues for Eclipse users because it was made in IntelliJ.

davidgraeff replied:
Because we have 3 official binding development IDEs now as seen in the official docs. If you say that we want to keep .project files I guess you are alright with also checking in .vscode directories and .idea files?

hakan42 replied:
ah, @wborn , does the classpath file as it is in the PR break everyone's workspace or would removing all .project and .classpaths overall break everything?

I just want to be sure to cause minimal disruption for everyone 😊

hakan42 replied:
Also, the PR is not really in a WIP state but actually usable.
I'll happily remove that label as soon as Jenkins manages to build things.

wborn replied:
The removal of all project files would break everyone's workspace. I'd add the removal to the OH3 TODO list so they are gone when the repo is shrunken (https://github.com/openhab/openhab2-addons/issues/5371). I usually don't commit Eclipse .project/.classpath files to Git on other projects. But they are here now because this repo has a different legacy.

hakan42 replied:
Added a nicer classpath file from a recently contributed binding (mail, to be precise)

davidgraeff commented:
You need to add this bundle to the bill of materials (bom/..) file as well :)

hakan42 commented:
rebuild

hakan42 commented:
rebuild...

hakan42 commented:
Build failures are not caused by the binding itself.

See https://github.com/openhab/openhab2-addons/pull/5582

hakan42 commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/handler/PushbulletHandler.java:
```diff
@@ -0,0 +1,216 @@
+        super(thing);
+    }
+
+    @Override public void handleCommand(ChannelUID channelUID, Command command) {
```
I am using settings exported from Eclipse and imported into my IntelliJ configuration.

I might install a secondary eclipse instance to just do the formatting before committing if the differences in formatting are too big.

hakan42 commented on bundles/org.openhab.binding.pushbullet/.project:
```diff
@@ -0,0 +1,33 @@
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.pde.PluginNature</nature>
```
fixed.

hakan42 replied:
Agreed. And fixed.

I still would love an IDE-neutral source tree though :)

hakan42 commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/handler/PushbulletHandler.java:
```diff
@@ -0,0 +1,216 @@
+        PushbulletConfiguration configuration = getConfigAs(PushbulletConfiguration.class);
+        logger.debug("CFG {}", configuration);
+        logger.debug("CFG Name  '{}'", configuration.getName());
+        logger.debug("CFG Token '{}'", configuration.getToken());
```
True. Fixed.

hakan42 replied:
Agreed. And removed the token log.

hakan42 commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/PushbulletBindingConstants.java:
```diff
@@ -0,0 +1,46 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.pushbullet;
```
Would that be all classes like handlers and action definitions and so on? Or just the constants? Could you point me to a known-to-be-good example bundle?

davidgraeff replied:
All classes. Constants are actually the exception, if they are used across multiple bundles (like with the MQTT main and sub bundles).

Hilbrand replied:
It should be:
```suggestion
package org.openhab.binding.pushbullet.internal;
```
The `internal` must be in the specific binding package not the other way around.

wborn replied:
That's right! :+1:

hakan42 replied:
o.k., done.

hakan42 commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/model/PushError.java:
```diff
@@ -0,0 +1,75 @@
+    @SerializedName("param")
+    private String param;
+
+    @SerializedName("cat")
```
It should. But I would love to keep them in the source so I can refactor the DTOs at will without fear of accidentally breaking the API access.

hakan42 commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/model/Push.java:
```diff
@@ -0,0 +1,86 @@
+    @SerializedName("type")
+    private String type;
+
+    @SerializedName("email")
```
see above at PushError for my reasoning to keep the annotations.

hakan42 commented:
@wborn , your suggested changes lead to DCO violations.

Do you want me to squash the commits and force-push or will you manually "accept" them? Also, is there a way to improve our DCO checker for those "Co-Authored-By" lines?

hakan42 commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/action/PushbulletActions.java:
```diff
@@ -0,0 +1,96 @@
+            return false;
+        }
+
+        // Ignore the SAT warning, we check for nullness just above this line.
```
Agreed, and fixed.

hakan42 commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/action/PushbulletActions.java:
```diff
@@ -0,0 +1,96 @@
+            return false;
+        }
+
+        // Ignore the SAT warning, we check for nullness just above this line.
```
agreed, and fixed.

wborn commented:
I'm OK with squashing all the commits @hakan42 to fix the sign offs.

It's not strictly necessary because we can also fix the sign offs when we squash them when merging this PR.

DCO failing on those commits seems to be a known issue: https://github.com/probot/dco/issues/102. 

Also note that we use an `Also-by:` line for additional authors instead of `Co-Authored-By:`, see [Sign your work](https://www.openhab.org/docs/developer/contributing.html#sign-your-work). 

If there are still pending or ongoing reviews it's usually better not to squash the commits so reviewers can check the changes.

hakan42 commented:
@wborn , @davidgraeff , I believe I adapted all your comments. Could you please give the PR another look? I'd love to have it merged for Milestone 2 😄

hakan42 commented:
Jenkins, please rebuild...

hakan42 commented:
Jenkins, please rebuild...

hakan42 commented:
rebased against the current master, and thus force-pushed...

@wborn , @davidgraeff , friendly ping ?

hakan42 commented:
Jenkins error caused while verifying bluez binding?

[ERROR] Failed to execute goal org.apache.karaf.tooling:karaf-maven-plugin:4.2.6:verify (karaf-feature-verification) on project org.openhab.binding.bluetooth.bluez: Feature resolution failed for [openhab-binding-bluetooth-bluez/2.5.0.SNAPSHOT]
[ERROR] Message: org.apache.karaf.features.internal.util.MultiException: Error:
[ERROR] Error downloading mvn:org.openhab/nrjavaserial/3.15.0.OH2

hakan42 commented:
Jenkins, please rebuild...

hakan42 commented:
@openhab/2-x-add-ons-maintainers , could I ask for a friendly "LGTM"? This binding is working in my productive environment for multiple weeks now.

I squashed the commits to get rid of a wrong commit email address (incorrectly used the work git configuration), and rebased against the current master...

martinvw commented on bundles/org.openhab.binding.pushbullet/README.md:
````diff
@@ -0,0 +1,162 @@
+end
+```
+
+
````
Could you maybe remove some of the newlines here

hakan42 replied:
Done.

martinvw commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java:
```diff
@@ -0,0 +1,218 @@
+        logger.debug("Start initializing!");
+        config = getConfigAs(PushbulletConfiguration.class);
+
+        // set the thing status to UNKNOWN temporarily and let the background task decide for the real status.
```
This comment conflicts with line 98

hakan42 replied:
True, was a leftover from the template code. Removed.

martinvw commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java:
```diff
@@ -0,0 +1,218 @@
+
+        logger.debug("sendPush is called for ");
+        logger.debug("Thing {}", thing);
+        logger.debug("Thing Label: '{}'", thing.getLabel());
```
Wouldn't this be resolved by the toString of the argument of the above logline

hakan42 replied:
In principle, yes. I just like being showed line-by-line debug information. Consider it a matter of taste please :)

martinvw commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java:
```diff
@@ -0,0 +1,218 @@
+
+        PushbulletConfiguration configuration = getConfigAs(PushbulletConfiguration.class);
+        logger.debug("CFG {}", configuration);
+        logger.debug("CFG Name  '{}'", configuration.getName());
```
Wouldn't this be resolved by the toString of the argument of the above logline

hakan42 replied:
In principle, yes. I just like being showed line-by-line debug information. Consider it a matter of taste please :)

martinvw commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java:
```diff
@@ -0,0 +1,218 @@
+            logger.debug("Packed Request: {}", request);
+
+            Properties headers = new Properties();
+            headers.put(HttpHeader.USER_AGENT, "openHAB / Pushbullet binding " + VERSION.toString());
```
I suppose this toString would not be needed

hakan42 replied:
True, removed.

martinvw commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java:
```diff
@@ -0,0 +1,218 @@
+                logger.warn("IO problems pushing note: {}", e.getMessage());
+            }
+
+        } catch (java.lang.Error e) {
```
Do not catch Error

hakan42 replied:
In this case, I have to catch Error to be able to obtain information about the HttpResponseException.

I could check for HttpResponseException, and if that is not the case, rethrow the original error. Would that be o.k. for you?

martinvw replied:
But who Throws an error, please not that errors are different branch from `Exception` it contains things like `OutOfMemmoryError`, `StackOverflow` etc

What exact error do you intend to catch and who threw it?

hakan42 replied:
Someone deep inside HttpUtil.executeUrl()... I'll try to provoke an exception so I can provide you with a full stack trace.

In the mean time, I'll check for the HttpResponseException (as mentioned above), and rethrow the exception if it did not contain what I was looking for. Would that be o.k. for you?

martinvw replied:
> Someone deep inside HttpUtil.executeUrl()... I'll try to provoke an exception so I can provide you with a full stack trace.
> 
> In the mean time, I'll check for the HttpResponseException (as mentioned above), and rethrow the exception if it did not contain what I was looking for. Would that be o.k. for you?

Exception would be 'fine' (and are indeed throw from HttpUtil) but you are catching the Error.

Please provide me with it, I would be surprised if its not an Exception but really an Error

hakan42 replied:
> Please provide me with it, I would be surprised if its not an `Exception` but really an `Error`

As luck would have it, I am getting "normal" JSON answers to an invalid token just now. Whereas, earlier, it was packaged in the Error. Maybe the API changed since I had written the 1.x action class.

I'll remove the complete error-catching part and wait whether it happens again, then we can talk about how to handle it.

martinvw replied:
> I'll remove the complete error-catching part and wait whether it happens again, then we can talk about how to handle it.

Great, just ping me when you are ready, we should be close to a merge then

hakan42 replied:
@martinvw , I did refactor as discussed, and Travis is happy too. I think Jenkins will also be whenever it gets around to build this branch.

Could you give the PushbulletHandler class another look please?

martinvw commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java:
```diff
@@ -0,0 +1,218 @@
+            logger.debug("Error in getResponseString: ", e);
+        }
+
+        return result;
```
Consider splitting up this method

hakan42 replied:
But how? logically, it is one operation. splitting it up would only serve to introduce multiple methods, but to no gain with regard to the logic that has to be performed.

martinvw replied:
I would say that it contains a lot of logic units, eg contructing the push message, or constructing and convert it to a string. Or performing the actual request.

hakan42 replied:
I'll refactor it later tonight, at the same time working on the error-catching part.

martinvw commented on bom/openhab-addons/pom.xml:
```diff
@@ -600,6 +600,11 @@
       <artifactId>org.openhab.binding.pulseaudio</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
```
@J-N-K / @maggu2810 / @openhab/2-x-add-ons-maintainers should a binding contain a feature.xml file now or only in special cases?

maggu2810 replied:
I did not follow the "a feature per addon" changes.
Perhaps this could be answered by @openhab/2-x-add-ons-maintainers

hakan42 replied:
As far as I can see, *all* bindings contain a feature.xml (looking at the HEAD of openhab2-addons repo). Therefore, I moved "my" feature.xml into the binding source tree as well.

hakan42 commented:
By the way, Jenkins failure is caused by an Out of Memory error while doing the karaf-feature-verification. Not directly caused by this binding, I hope :)

hakan42 commented:
Jenkins, please rebuild

martinvw commented on bundles/pom.xml:
```diff
@@ -156,6 +156,7 @@
     <module>org.openhab.binding.plugwise</module>
     <module>org.openhab.binding.powermax</module>
     <module>org.openhab.binding.pulseaudio</module>
+    <module>org.openhab.binding.pushbullet</module>
```
Could you add the binding to the codeowners

hakan42 replied:
The first file in the "changes" tab shows the addition to CODEOWNERS.

I'll wait for your merge, and if does not show up in the HEAD , I'll create an additional PR as suggested

martinvw commented on bundles/org.openhab.binding.pushbullet/src/main/java/org/openhab/binding/pushbullet/internal/handler/PushbulletHandler.java:
```diff
@@ -0,0 +1,229 @@
+            if ((null != response) && (null == response.getPushError())) {
+                result = true;
+            }
+        }
```
Formatting is a little bit a-typical

hakan42 replied:
Maybe caused by my IntelliJ installation. It should have applied the Eclipse formatting rules though.

Can I fix that in an upcoming PR when I touch this code again?

martinvw commented:
> Looks good except those multiple debug lines (remember that each log line is executed even if not at the debug level. Meaning loading the arguments to registers, initiate a call stack, jump to the method, call the framework to determine the logger level we are in, aborting)

Agree on that, it would be nice if you be logging less next time or combine it in a lot less statements with the proper placeholders etc.

hakan42 commented:
Thank you so much guys. And off I go, working on the next binding :)
